modules/platforms/cpp/ignite/client/detail/transaction/transaction_impl.h

```diff
@@ -0,0 +1,152 @@
+    enum class state {
+        OPEN,
+
+        COMMITTED,
+
+        ROLLED_BACK
+    };
+
+    // Deleted
+    transaction_impl() = delete;
+    transaction_impl(transaction_impl &&) noexcept = delete;
+    transaction_impl(const transaction_impl &) = delete;
+    transaction_impl &operator=(transaction_impl &&) noexcept = delete;
+    transaction_impl &operator=(const transaction_impl &) = delete;
+
+    /**
+     * Constructor.
+     *
+     * @param id Transaction ID.
+     * @param connection Connection.
+     */
+    explicit transaction_impl(std::int64_t id, std::shared_ptr<node_connection> connection)
+        : m_id(id)
+        , m_state(state::OPEN)
+        , m_connection(std::move(connection)) {}
+
+    ~transaction_impl() {
+        state old = state::OPEN;
+        if (m_state.compare_exchange_strong(old, state::ROLLED_BACK, std::memory_order_relaxed))
+        {
+            finish(false, [](auto){});
+        }
+    }
+
+    /**
+     * Commits the transaction asynchronously.
+     *
+     * @param callback Callback to be called upon asynchronous operation completion.
+     */
+    void commit_async(ignite_callback<void> callback) {
+        set_state(state::COMMITTED);
+
+        finish(true, std::move(callback));
+    }
+
+    /**
+     * Rollbacks the transaction asynchronously.
+     *
+     * @param callback Callback to be called upon asynchronous operation completion.
+     */
+    void rollback_async(ignite_callback<void> callback) {
+        set_state(state::ROLLED_BACK);
+
+        finish(false, std::move(callback));
+    }
+
+    /**
+     * Get transaction ID.
+     *
+     * @return Transaction ID.
+     */
+    [[nodiscard]] std::int64_t get_id() const {
+        return m_id;
+    }
+
+    /**
+     * Get connection.
+     *
+     * @return Connection.
+     */
+    [[nodiscard]] std::shared_ptr<node_connection> get_connection() const {
+        return m_connection;
+    }
+
+private:
+    /**
+     * Perform operation.
+     *
+     * @param commit Flag indicating should transaction be committed or rolled back.
+     * @param callback Callback to be called upon asynchronous operation completion.
+     */
+    void finish(bool commit, ignite_callback<void> callback) {
+        auto writer_func = [id = m_id](protocol::writer &writer) { writer.write(id); };
+
+        m_connection->perform_request_wr<void>(
+            commit ? client_operation::TX_COMMIT : client_operation::TX_ROLLBACK, writer_func, std::move(callback));
+    }
+
+    /**
+     * Set state.
+     *
+     * @param st State.
+     */
+    void set_state(state st) {
+        state old = state::OPEN;
+        auto opened = m_state.compare_exchange_strong(old, st, std::memory_order_relaxed);
+        if (opened)
+            return;
+
+        auto message = old == state::COMMITTED
+            ? "Transaction is already committed."
+            : "Transaction is already rolled back.";
+
+        throw ignite_error(message);
```

Comment:
This behavior will be changed soon on the Java side: https://issues.apache.org/jira/browse/IGNITE-18324
As I understand, commit/rollback will be ignored if a tx is already committed or rolled back.

Reply:
Looks weird, but OK, let's change it too.

Reply:
We can probably do it later after Java-side changes are finalized and merged, to ensure consistent behavior.